plugin/rename.py

```diff
@@ -1,61 +1,58 @@
 from .core.url import parse_uri
 from .core.views import first_selection_region
 from .core.views import get_line
 from .core.views import range_to_region
 from .core.views import text_document_position_params
-import functools
+from functools import partial
 import os
 import sublime
 import sublime_plugin
 
-
-class RenameSymbolInputHandler(sublime_plugin.TextInputHandler):
-    def want_event(self) -> bool:
-        return False
-
-    def __init__(self, view: sublime.View, placeholder: str) -> None:
-        self.view = view
-        self._placeholder = placeholder
-
-    def name(self) -> str:
-        return "new_name"
-
-    def placeholder(self) -> str:
-        return self._placeholder
-
-    def initial_text(self) -> str:
-        return self.placeholder()
-
-    def validate(self, name: str) -> bool:
-        return len(name) > 0
-
+def is_range_response(result: PrepareRenameResult) -> TypeGuard[Range]:
+    return 'start' in result
+
+# The flow of this command is fairly complicated so it deserves some documentation.
+#
+# When "LSP: Rename" is triggered from the Command Palette, the flow can go one of two ways:
+#
+# 1. Session doesn't have support for "prepareProvider":
+#  - input() gets called with empty "args" - returns an instance of "RenameSymbolInputHandler"
+#  - input overlay triggered
+#  - user enters new name and confirms
+#  - run() gets called with "new_name" argument
+#  - rename is performed
+#
+# 2. Session has support for "prepareProvider":
+#  - input() gets called with empty "args" - returns None
+#  - run() gets called with no arguments
+#  - "prepare" request is triggered on the session
+#  - based on the "prepare" response, the "placeholder" value is computed
+#  - "lsp_symbol_rename" command is re-triggered with computed "placeholder" argument
+#  - run() gets called with "placeholder" argument set
+#  - run() manually throws a TypeError
+#  - input() gets called with "placeholder" argument set - returns an instance of "RenameSymbolInputHandler"
+#  - input overlay triggered
+#  - user enters new name and confirms
+#  - run() gets called with "new_name" argument
+#  - rename is performed
+#
+# Note how triggering the command programmatically triggers run() first while when triggering the command from
+# the Command Palette the input() gets called first.
 
 class LspSymbolRenameCommand(LspTextCommand):
 
     capability = 'renameProvider'
 
-    # mypy: Signature of "is_enabled" incompatible with supertype "LspTextCommand"
-    def is_enabled(  # type: ignore
-        self,
-        new_name: str = "",
-        placeholder: str = "",
-        position: Optional[int] = None,
-        event: Optional[dict] = None,
-        point: Optional[int] = None
-    ) -> bool:
-        if self.best_session("renameProvider.prepareProvider"):
-            # The language server will tell us if the selection is on a valid token.
-            return True
-        return super().is_enabled(event, point)
```

Comment:
This was redundant and it's enough to use base implementation.
When this code was introduced it looked a bit differently and it made sense then.